### PR TITLE
Image moments function

### PR DESCRIPTION
**Pull request by syurkevi**
Implements m00, m01, m10, m11 image moments for all backends. Addresses #1325.

### CODE REVIEW

**Comment by shehzan10** [comment]
Split adding the test and updating test/data into a new commit. You can force push after that.

File: include/af/defines.h [hunk 1]
```diff
@@ -357,6 +357,15 @@ typedef enum {
 } af_image_format;
 #endif
 
+#if AF_API_VERSION >=34
+typedef enum {
+    M00 = 0,
```
Comment (shehzan10) [inline include/af/defines.h]: Changes:
MXY ->`AF_MOMENT_XY`.
af_moment -> af_moment_type

Also you missed adding `typedef af_moment_type af::momentType` in the `namespace af` later in the file.

File: include/af/image.h [hunk 2]
```diff
@@ -677,6 +677,32 @@ AFAPI array ycbcr2rgb(const array& in, const YCCStd standard=AF_YCC_601);
...
+/**
+   C++ Interface for calculating an image moment
+
+   \param[in]  in is the input image
+   \param[moment] is the moment to calculate
+   \return      the value of the moment
+
+   \ingroup image_func_moments
+ */
+template<typename T> T moments(const array& in, const af_moment moment);
```
Comment (shehzan10) [inline include/af/image.h]: Did you intentionally leave out the c-api counterpart for this function? Use a function name like `af_moments_all`

File: src/backend/cpu/image.cpp [hunk 3]
```diff
@@ -40,8 +42,44 @@ void copy_image(const Array<T> &in, const fg::Image* image)
     CheckGL("In CopyArrayToPBO");
 }
 
+template<typename T>
+Array<float> moments(const Array<T> &in, const af_moment moment)
```
Comment (shehzan10) [inline src/backend/cpu/image.cpp]: Put this in a new file called moments.cpp. Image.cpp is for the image() function. Don't mix them up. Also, if graphics is disabled, this will cause error.

File: src/backend/cuda/image.cu [hunk 4]
```diff
@@ -57,8 +58,43 @@ void copy_image(const Array<T> &in, const fg::Image* image)
     }
 }
 
-#define INSTANTIATE(T)      \
-    template void copy_image<T>(const Array<T> &in, const fg::Image* image);
+template<typename T>
+Array<float> moments(const Array<T> &in, const af_moment moment)
```
Comment (shehzan10) [inline src/backend/cuda/image.cu]: Same as CPU, move it into a new file.

File: test/image.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
+/*******************************************************
```
Comment (shehzan10) [inline test/image.cpp]: Rename this file to moment.cpp

File: include/af/defines.h (hunk 1 above)
Comment (shehzan10) [inline include/af/defines.h]: Another point here is do you want to use moment or moments? Most places I've seen, you are using moments.

File: include/af/image.h (hunk 2 above)
Comment (pavanky) [inline include/af/image.h]: `af_image_moment` (cases where you pass single image and expect single output) and `af_moments` (where you can have multiple images) might be appropriate I think

File: src/backend/cpu/kernel/moments.hpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                 dim_t const idx, dim_t const idy, dim_t const idz, dim_t const idw)
+    {
+        dim_t mId = idy * istrides[1] + idx;
+        if(pBatch) {
+            mId += idw * istrides[3] + idz * istrides[2];
+        }
+
+        return idx * idy * in[mId];
+    }
+};
```
Comment (pavanky) [inline src/backend/cpu/kernel/moments.hpp]: Don't pass in strides. Calculate idx, idy, idz, idw outside and pass those along aling side mId.

File: src/backend/cpu/kernel/moments.hpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    af::dim4  const odims    = output.dims();
+
+    moments_op<T, Method> op;
+    bool pBatch = !(idims[2] == 1 && idims[3] == 1);
+    bool tDim   = (idims[3] != 1);
+    bool zDim   = (idims[2] != 1);
+
+    for(dim_t w = 0; w < idims[3]; w++) {
+        T val = scalar<T>(0);
+        for(dim_t z = 0; z < idims[2]; z++) {
```
Comment (pavanky) [inline src/backend/cpu/kernel/moments.hpp]: Move `val = scalar<T>(0)` here. You don't need it in two places.

File: src/backend/cpu/kernel/moments.hpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            for(dim_t y = 0; y < idims[1]; y++) {
+                for(dim_t x = 0; x < idims[0]; x++) {
+                    val += op(in, idims, iElems, istrides, pBatch, x, y, z, w);
+                }
+            }
+            output.get()[w * odims[0] + z] = (float)val;
```
Comment (pavanky) [inline src/backend/cpu/kernel/moments.hpp]: Don't do this. Store the pointer some place ahead of time.

File: src/api/cpp/image.cpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
```
Comment (pavanky) [inline src/api/cpp/image.cpp]: 2016 everywhere please.

File: src/backend/cuda/kernel/moments.hpp [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        return idx * idy * in.ptr[mId];
+    }
+
+    template<typename T, af_moment moment>
+    __global__
+    void moments_kernel(CParam<float> out, CParam<T> in,
+                  const dim_t blocksMatX, const bool pBatch)
+    {
+        const dim_t idw = blockIdx.y / in.dims[2];
+        const dim_t idz = blockIdx.y - idw * in.dims[2];
```
Comment (pavanky) [inline src/backend/cuda/kernel/moments.hpp]: you can check for `pBatch` here and offset the last two dimensions before calling the functors.

File: src/backend/cuda/kernel/moments.hpp [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        if(pBatch) {
+            mId += idw * in.strides[3] + idz * in.strides[2];
+        }
+
+        return idx * idy * in.ptr[mId];
+    }
+
+    template<typename T, af_moment moment>
+    __global__
+    void moments_kernel(CParam<float> out, CParam<T> in,
```
Comment (pavanky) [inline src/backend/cuda/kernel/moments.hpp]: Output has to be `Param<float>`

File: src/backend/cpu/image.cpp [hunk 12]
```diff
@@ -40,8 +42,44 @@ void copy_image(const Array<T> &in, const fg::Image* image)
...
+{
+    dim4 odims, idims = in.dims();
+    odims[0] = odims[1] = odims[2] = odims[3] = 1;
+    if(idims[2] != 1) {
+        odims[0] = idims[2];
+    }
+    if(idims[3] != 1) {
+        odims[1] = idims[3];
+    }
```
Comment (pavanky) [inline src/backend/cpu/image.cpp]: Seriously?

File: src/backend/opencl/kernel/moments.cl [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+    if(idx >= in.dims[0] ||
+       idy >= in.dims[1] ||
+       idz >= in.dims[2] ||
+       idw >= in.dims[3])
+        return;
+
+    __local T wkg_moment_sum;
+    wkg_moment_sum = 0;
+    barrier(CLK_LOCAL_MEM_FENCE);
```
Comment (pavanky) [inline src/backend/opencl/kernel/moments.cl]: Why do you have a barrier without setting anything to local memory ?

Comment (pavanky) [inline src/backend/opencl/kernel/moments.cl]: nevermind.

File: src/backend/opencl/kernel/moments.hpp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                    momentsKernels[device] = new Kernel(*momentsProgs[device], "moments_kernel");
+                });
+
+
+                auto momentsp = make_kernel<Buffer, const KParam, const Buffer, const KParam,
+                                       const int, const int>
+                                      (*momentsKernels[device]);
+
+                NDRange local(THREADS, 1, 1);
+                dim_t blocksPerMat = divup(in.info.dims[0], local[0]);
```
Comment (pavanky) [inline src/backend/opencl/kernel/moments.hpp]: The variable name is a misnomer. it should technically be blocksPerRow / blocksPerX

File: src/backend/cuda/kernel/moments.hpp [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+#include <debug_cuda.hpp>
+
+namespace cuda
+{
+namespace kernel
+{
+
+    // Kernel Launch Config Values
+    static const int THREADS = 128;
+
```
Comment (pavanky) [inline src/backend/cuda/kernel/moments.hpp]: Why 128 ?

Comment (syurkevi) [inline src/backend/cuda/kernel/moments.hpp]: (256) 1.68843s vs (128) 1.20581s 
on a 1000 x 1000 x 200 input

File: include/af/image.h [hunk 16]
```diff
@@ -1348,6 +1348,7 @@ extern "C" {
     */
     AFAPI af_err af_rgb2ycbcr(af_array* out, const af_array in, const af_ycc_std standard);
 #endif
+
```
Comment (shehzan10) [inline include/af/image.h]: why?

File: include/af/moments.h [hunk 17]
```diff
@@ -0,0 +1,83 @@
+/*******************************************************
```
Comment (shehzan10) [inline include/af/moments.h]: Make only the functions listed in this file part of `image.h`. We should avoid creating new header files unless the functions are of a completely new domain.

File: include/arrayfire.h [hunk 18]
```diff
@@ -268,6 +268,7 @@
 #include "af/image.h"
 #include "af/index.h"
 #include "af/lapack.h"
+#include "af/moments.h"
```
Comment (shehzan10) [inline include/arrayfire.h]: Once you remove moments.h, you can remove this too

File: src/api/c/moments.cpp [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    Array<T> output = castArray<T>(tarr);
+
+    *out = getHandle<T>(output);
+}
+
+af_err af_moments(af_array *out, const af_array in, const af_moment_type moment)
+{
+    try {
+        const ArrayInfo in_info = getInfo(in);
+        af_dtype type = in_info.getType();
```
Comment (shehzan10) [inline src/api/c/moments.cpp]: Doesn't this function need more checks like the number of dimensions (2D/3D etc)?

File: src/api/c/moments.cpp [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+    dim_t elems;
+    af_get_elements(&elems, moments);
+    T *h_ptr = new T[elems];
+    af_get_data_ptr((void *)h_ptr, moments);
+
+    *out = (double)h_ptr[0];
+    delete[] h_ptr;
+}
+
+af_err af_moment(double *out, const af_array in, const af_moment_type moment)
```
Comment (shehzan10) [inline src/api/c/moments.cpp]: You cannot overload C-API. Call this af_moment_all.

File: src/api/cpp/moments.cpp [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+        return (T)out;                                      \
+    }
+
+
+INSTANTIATE_REAL(float)
+INSTANTIATE_REAL(double)
+INSTANTIATE_REAL(int)
+INSTANTIATE_REAL(unsigned)
+INSTANTIATE_REAL(long long)
+INSTANTIATE_REAL(unsigned long long)
```
Comment (shehzan10) [inline src/api/cpp/moments.cpp]: Use intl for long long and uintl for unsigned long long.

File: src/api/unified/moments.cpp [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+
+#include <af/array.h>
+#include <af/moments.h>
+#include "symbol_manager.hpp"
+
+af_err af_moments(af_array* out, const af_array in, const af_moment_type moment)
+{
+    CHECK_ARRAYS(in);
+    return CALL(out, in, moment);
+}
```
Comment (shehzan10) [inline src/api/unified/moments.cpp]: You should defined the moment function with the double return argument here too.

File: src/backend/cpu/image.cpp [hunk 23]
```diff
@@ -12,11 +12,13 @@
 
 #if defined (WITH_GRAPHICS)
 
+#include <af/defines.h>
 #include <Array.hpp>
 #include <image.hpp>
 #include <err_cpu.hpp>
 #include <graphics_common.hpp>
 #include <platform.hpp>
+#include <kernel/moments.hpp>
```
Comment (shehzan10) [inline src/backend/cpu/image.cpp]: Don't leave residual code.

File: src/backend/cpu/image.hpp [hunk 24]
```diff
@@ -11,6 +11,7 @@
 
 #include <Array.hpp>
 #include <graphics_common.hpp>
+#include <af/defines.h>
```
Comment (shehzan10) [inline src/backend/cpu/image.hpp]: You probably don't need this here. defines.h is included from Array.h. You also don't need graphics_common.hpp.

File: src/backend/cpu/kernel/moments.hpp [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+struct moments_op<T, AF_MOMENT_M11>
+{
+    T operator()(T const * const in, dim_t mId, dim_t const idx, dim_t const idy)
+    {
+        return idx * idy * in[mId];
+    }
+};
+
+template<typename T, af_moment_type Method>
+void moments(Array<float> &output, Array<T> const input)
```
Comment (shehzan10) [inline src/backend/cpu/kernel/moments.hpp]: Use const Array<T> &input

File: src/backend/cpu/kernel/moments.hpp [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+    for(dim_t w = 0; w < idims[3]; w++) {
+        for(dim_t z = 0; z < idims[2]; z++) {
+            T val = scalar<T>(0);
+            for(dim_t y = 0; y < idims[1]; y++) {
+                for(dim_t x = 0; x < idims[0]; x++) {
+                    val += op(in, mId, x, y);
+                    mId++;
+                }
+            }
+            out[w * odims[0] + z] = (float)val;
```
Comment (shehzan10) [inline src/backend/cpu/kernel/moments.hpp]: When calculating an index, always use strides, not dims.

File: src/backend/cpu/moments.cpp [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+ ********************************************************/
+
+#include <af/defines.h>
+#include <Array.hpp>
+#include <err_cpu.hpp>
+#include <platform.hpp>
+#include <kernel/moments.hpp>
+#include <queue.hpp>
+
+using af::dim4;
```
Comment (shehzan10) [inline src/backend/cpu/moments.cpp]: Put this inside `namespace cpu`.

File: src/backend/cpu/moments.cpp [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+            break;
+        default:  break;
+    }
+    getQueue().sync();
+    return out;
+}
+
+
+#define INSTANTIATE(T)  \
+    template Array<float> moments<T>(const Array<T> &in, const af_moment_type moment);
```
Comment (shehzan10) [inline src/backend/cpu/moments.cpp]: Shouldn't it be the case that if input is double, the return type should be Array<double>? Or is it safe to use float?

File: src/backend/cuda/image.cu [hunk 29]
```diff
@@ -57,7 +58,7 @@ void copy_image(const Array<T> &in, const fg::Image* image)
     }
 }
 
-#define INSTANTIATE(T)      \
+#define INSTANTIATE(T)                                                          \
```
Comment (shehzan10) [inline src/backend/cuda/image.cu]: Residual code.

File: src/backend/cuda/kernel/moments.hpp [hunk 30]
```diff
@@ -0,0 +1,120 @@
...
+                break;
+            case AF_MOMENT_M11:
+                atomicAdd(&blk_moment_sum, (float)moments_m11(mId, idx, idy, in));
+                break;
+            default:
+                break;
+        }
+        __syncthreads();
+
+        float *offset = const_cast<float *>(out.ptr + (idw * out.dims[0] + idz));
```
Comment (shehzan10) [inline src/backend/cuda/kernel/moments.hpp]: Use strides, not dims.

File: src/backend/cuda/kernel/moments.hpp [hunk 31]
```diff
@@ -0,0 +1,120 @@
...
+        dim3 threads(THREADS, 1, 1);
+        dim_t blocksMatX = divup(in.dims[0], threads.x);
+        dim3 blocks(blocksMatX * in.dims[1], in.dims[2] * in.dims[3]);
+
+        bool pBatch = !(in.dims[2] == 1 && in.dims[3] == 1);
+
+        CUDA_LAUNCH((moments_kernel<T, moment>), blocks, threads,
+                     out, in, blocksMatX, pBatch);
+        POST_LAUNCH_CHECK();
+        CUDA_CHECK(cudaDeviceSynchronize());
```
Comment (shehzan10) [inline src/backend/cuda/kernel/moments.hpp]: Nope

File: src/backend/cuda/moments.cu [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <Array.hpp>
+#include <err_cuda.hpp>
+#include <debug_cuda.hpp>
+#include <interopManager.hpp>
```
Comment (shehzan10) [inline src/backend/cuda/moments.cu]: Don't need this here.

File: src/backend/cuda/moments.cu [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+#include <kernel/moments.hpp>
+
+using af::dim4;
```
Comment (shehzan10) [inline src/backend/cuda/moments.cu]: Inside namespace cuda

File: src/backend/cuda/moments.cu [hunk 34]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    if(idims[3] != 1) {
+        odims[1] = idims[3];
+    }
+
+    in.eval();
+    Array<float> out = createValueArray<float>(odims, 0.f);
+
+    switch(moment) {
+        case AF_MOMENT_M00:
```
Comment (shehzan10) [inline src/backend/cuda/moments.cu]: Make these case statements into a single line each
`case M00: function; break;`

File: src/backend/opencl/image.cpp [hunk 35]
```diff
@@ -62,7 +63,7 @@ void copy_image(const Array<T> &in, const fg::Image* image)
     }
 }
 
-#define INSTANTIATE(T)      \
+#define INSTANTIATE(T)                                                              \
```
Comment (shehzan10) [inline src/backend/opencl/image.cpp]: Undo the changes in this file

File: src/backend/opencl/kernel/moments.cl [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+
+    if(idx >= in.dims[0] ||
+       idy >= in.dims[1] ||
+       idz >= in.dims[2] ||
+       idw >= in.dims[3])
+        return;
+
+    __local float wkg_moment_sum;
+    wkg_moment_sum = 0;
+    barrier(CLK_LOCAL_MEM_FENCE);
```
Comment (shehzan10) [inline src/backend/opencl/kernel/moments.cl]: I don't think is barrier is necessary. You are practically guaranteeing that there will be a block-size atomic conflict. Thank is all the threads in the blocks will want to atomic add at the exact same step.

File: src/backend/opencl/moments.cpp [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <Array.hpp>
+#include <kernel/moments.hpp>
+#include <err_opencl.hpp>
+#include <debug_opencl.hpp>
+#include <stdexcept>
+#include <vector>
```
Comment (shehzan10) [inline src/backend/opencl/moments.cpp]: Do you need vector and stdexcept here?

File: src/backend/opencl/moments.cpp [hunk 38]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    if(idims[3] != 1) {
+        odims[1] = idims[3];
+    }
+
+    in.eval();
+    Array<float> out = createValueArray<float>(odims, 0.f);
+
+    switch(moment) {
+        case AF_MOMENT_M00:
```
Comment (shehzan10) [inline src/backend/opencl/moments.cpp]: Make these single lines.

**Comment by shehzan10** [comment]
build arrayfire linux ci

File: src/backend/cpu/moments.cpp (hunk 28 above)
Comment (pavanky) [inline src/backend/cpu/moments.cpp]: I think we can safely return floats for moments

File: src/backend/opencl/kernel/moments.cl (hunk 36 above)
Comment (syurkevi) [inline src/backend/opencl/kernel/moments.cl]: Don't I need to initialize wkg_moment_sum? If there is no barrier, some thread may overwrite the intermediate sum with 0...

Comment (pavanky) [inline src/backend/opencl/kernel/moments.cl]: No this is necessary. Ideally it should be `if (get_local_size(0) == 0) wkg_moment_sum = 0;` but this works as well.

If you don't do this, some of the threads in the workgroup might perform an atomicadd before the other threads reach this part of the code.

File: include/af/image.h [hunk 39]
```diff
@@ -677,6 +677,32 @@ AFAPI array ycbcr2rgb(const array& in, const YCCStd standard=AF_YCC_601);
...
+/**
+   C++ Interface for calculating an image moment
+
+   \param[in]  in is the input image
+   \param[moment] is the moment to calculate
+   \return      the value of the moment
+
+   \ingroup image_func_moments
+ */
+template<typename T> T moment(const array& in, const af_moment_type moment);
```
Comment (9prady9) [inline include/af/image.h]: @syurkevi There might be a similar comment down somewhere, add a typedef such as `typedef af_moment_type MomentType` under namespace af and use that here.

Comment (shehzan10) [inline include/af/image.h]: Yes use the typedef-ed version in the C++ API
Style: split into 2 lines
```
template <typename T>\
T moment....
```

File: include/af/defines.h [hunk 40]
```diff
@@ -357,6 +357,16 @@ typedef enum {
...
 #endif
 
+#if AF_API_VERSION >=34
+typedef enum {
+    AF_MOMENT_M00 = 1,
+    AF_MOMENT_M01 = 2,
+    AF_MOMENT_M10 = 4,
+    AF_MOMENT_M11 = 8,
+    AF_MOMENT_FIRST_ORDER = 0xF
+} af_moment_type;
```
Comment (pavanky) [inline include/af/defines.h]: Can you make this `AF_MOMENT_FIRST_ORDER = AF_MOMENT_M00 | AF_MOMENT_M01 | AF_MOMENT_M10 | AF_MOMENT_M11` ?

File: include/af/image.h [hunk 41]
```diff
@@ -677,6 +677,32 @@ AFAPI array ycbcr2rgb(const array& in, const YCCStd standard=AF_YCC_601);
...
+   C++ Interface for calculating an image moment
+
+   \param[in]  in is the input image
+   \param[out] out is a pointer to the outputted moment(s)
+   \param[moment] is the moment(s) to calculate
+
+   \ingroup image_func_moments
+ */
+AFAPI void moments(double* out, const array& in, const af_moment_type moment);
+#endif
```
Comment (pavanky) [inline include/af/image.h]: use `momentType` here. And set default value to be  `AF_MOMENT_FIRST_ORDER`

File: include/af/image.h [hunk 42]
```diff
@@ -677,6 +677,32 @@ AFAPI array ycbcr2rgb(const array& in, const YCCStd standard=AF_YCC_601);
...
+
+   \param[in]  in contains the input image(s)
+   \param[moment] is the moment to calculate
+   \return array containing the requested moment of each image
+
+   \ingroup image_func_moments
+ */
+AFAPI array moments(const array& in, const af_moment_type moment);
+#endif
+
```
Comment (pavanky) [inline include/af/image.h]: use `momentType` here. And set default value to be  `AF_MOMENT_FIRST_ORDER`

File: include/af/image.h [hunk 43]
```diff
@@ -1348,6 +1374,37 @@ extern "C" {
...
+       \ingroup image_func_moments
+    */
+    AFAPI af_err af_moments(af_array *out, const af_array in, const af_moment_type moment);
+#endif
+
+#if AF_API_VERSION >= 34
+    /**
+       C Interface for calculating image moment(s) of a single image
+
+       \param[out] out is  a pointer to the outputted moment(s)
```
Comment (pavanky) [inline include/af/image.h]: Change the phrasing to be clearer. Also mention user needs to have adequate space allocated.

File: include/af/image.h [hunk 44]
```diff
@@ -677,6 +677,32 @@ AFAPI array ycbcr2rgb(const array& in, const YCCStd standard=AF_YCC_601);
 AFAPI array rgb2ycbcr(const array& in, const YCCStd standard=AF_YCC_601);
 #endif
 
+#if AF_API_VERSION >= 34
+/**
+   C++ Interface for calculating an image moment
+
+   \param[in]  in is the input image
+   \param[out] out is a pointer to the outputted moment(s)
+   \param[moment] is the moment(s) to calculate
```
Comment (pavanky) [inline include/af/image.h]: Should be `param[in] moment is moment(s) to calculate`.

File: include/af/image.h [hunk 45]
```diff
@@ -677,6 +677,32 @@ AFAPI array ycbcr2rgb(const array& in, const YCCStd standard=AF_YCC_601);
...
+ */
+AFAPI void moments(double* out, const array& in, const af_moment_type moment);
+#endif
+
+#if AF_API_VERSION >= 34
+/**
+   C++ Interface for calculating image moments
+
+   \param[in]  in contains the input image(s)
+   \param[moment] is the moment to calculate
```
Comment (pavanky) [inline include/af/image.h]: Should be `param[in] moment is moment(s) to calculate`.

File: include/af/image.h [hunk 46]
```diff
@@ -1348,6 +1374,37 @@ extern "C" {
...
     AFAPI af_err af_rgb2ycbcr(af_array* out, const af_array in, const af_ycc_std standard);
 #endif
+
+#if AF_API_VERSION >= 34
+    /**
+       C Interface for finding image moments
+
+       \param[out] out is an array containing the calculated moments
+       \param[in]  in is an array of image(s)
+       \param[moment] is the moment(s) to calculate
```
Comment (pavanky) [inline include/af/image.h]: Should be `param[in] moment is moment(s) to calculate`.

File: include/af/image.h [hunk 47]
```diff
@@ -1348,6 +1374,37 @@ extern "C" {
...
+    AFAPI af_err af_moments(af_array *out, const af_array in, const af_moment_type moment);
+#endif
+
+#if AF_API_VERSION >= 34
+    /**
+       C Interface for calculating image moment(s) of a single image
+
+       \param[out] out is  a pointer to the outputted moment(s)
+       \param[in] in is the input image
+       \param[moment] is the moment(s) to calculate
```
Comment (pavanky) [inline include/af/image.h]: Should be `param[in] moment is moment(s) to calculate`.

File: src/backend/cpu/kernel/moments.hpp [hunk 48]
```diff
@@ -0,0 +1,67 @@
...
+                        out[w * ostrides[3] + z * ostrides[2] + m_off] += x * val;
+                        m_off++;
+                    }
+                    if((moment & AF_MOMENT_M10) > 0) {
+                        out[w * ostrides[3] + z * ostrides[2] + m_off] += y * val;
+                        m_off++;
+                    }
+                    if((moment & AF_MOMENT_M11) > 0) {
+                        out[w * ostrides[3] + z * ostrides[2] + m_off] += x * y * val;
+                        m_off++;
```
Comment (pavanky) [inline src/backend/cpu/kernel/moments.hpp]: calculate `w * ostrides[3] + z * ostrides[2]` before hand.

File: src/backend/cuda/kernel/moments.hpp [hunk 49]
```diff
@@ -0,0 +1,94 @@
...
+    void moments_kernel(CParam<float> out, CParam<T> in, af_moment_type moment,
+                  const dim_t blocksMatX, const bool pBatch)
+    {
+        const dim_t idw = blockIdx.y / in.dims[2];
+        const dim_t idz = blockIdx.y - idw * in.dims[2];
+
+        const dim_t idy = blockIdx.x;
+        dim_t idx = threadIdx.x;
+
+        __shared__ float blk_moment_sum[4];
```
Comment (pavanky) [inline src/backend/cuda/kernel/moments.hpp]: This shouldn't be 4.

File: src/backend/cuda/kernel/moments.hpp [hunk 50]
```diff
@@ -0,0 +1,94 @@
...
+
+        for(unsigned i=0; i<blocksMatX; ++i) {
+            dim_t mId = idy * in.strides[1] + idx;
+            if(pBatch) {
+                mId += idw * in.strides[3] + idz * in.strides[2];
+            }
+
+            if (idx >= in.dims[0] || idy >= in.dims[1] ||
+                idz >= in.dims[2] || idw >= in.dims[3] )
+                break;
```
Comment (pavanky) [inline src/backend/cuda/kernel/moments.hpp]: You can check for the `id{y,z,w}` conditions before entering the loop and only check for `idx` in here.

File: src/backend/cuda/kernel/moments.hpp [hunk 51]
```diff
@@ -0,0 +1,94 @@
...
+            blk_moment_sum[threadIdx.x] = 0.f;
+        }
+        __syncthreads();
+
+        for(unsigned i=0; i<blocksMatX; ++i) {
+            dim_t mId = idy * in.strides[1] + idx;
+            if(pBatch) {
+                mId += idw * in.strides[3] + idz * in.strides[2];
+            }
+
```
Comment (pavanky) [inline src/backend/cuda/kernel/moments.hpp]: Move this outside the for loop

**Comment by pavanky** [comment]
build arrayfire linux ci
build arrayfire osx ci
build arrayfire windows ci

**Comment by pavanky** [comment]
@syurkevi It looks like you did not update the submodule tag. Can you update it and push it?

**Comment by pavanky** [comment]
build arrayfire linux ci
build arrayfire osx ci

**Comment by pavanky** [comment]
@syurkevi can you add `if (noDoubleTests<T>()) return;` to the tests? Running double tests on machines that do not have double precision will cause problems.

File: test/moments.cpp [hunk 52]
```diff
@@ -87,6 +89,8 @@ void momentsTest(string pTestFile)
 
 void momentsOnImageTest(string pTestFile, string pImageFile, bool isColor)
 {
+    if (noDoubleTests<double>()) return;
+    if (noImageIOTests()) return;
```
Comment (pavanky) [inline test/moments.cpp]: Is this test really using doubles at any point?

**Comment by pavanky** [comment]
build arrayfire linux ci
build arrayfire osx ci

**Comment by pavanky** [comment]
build arrayfire windows ci
